### PR TITLE
Retrieve Account UUID From User Claims

### PR DESCRIPTION
**Pull request by refs**
For more context: https://owncloud.github.io/ocis/request-flow/

OIDC Middleware can make use of `uuidFromClaims` to trade claims.Email for an account's `UUID`. For this, a general purpose cache was added that caches on a per-request basis, meaning whenever the request parameters match a set of keys, the cached value is returned, saving a round trip to the accounts service **that otherwise would happen in every single request.**

## Preconditions

- proxy config requires an `OIDC` configuration.

### CODE REVIEW

File: pkg/server/http/server.go [hunk 1]
```diff
@@ -48,10 +50,8 @@ func Server(opts ...Option) (svc.Service, error) {
 		svc.Address(options.Config.HTTP.Addr),
 		svc.Context(options.Context),
 		svc.Flags(options.Flags...),
-		svc.Handler(applyMiddlewares(
-			options.Handler,
-			options.Middlewares...,
-		),
+		svc.Handler(
+			applyMiddlewares(options.Handler, options.Middlewares...)
```
Comment (kulmann) [inline pkg/server/http/server.go]: misses a trailing comma

Comment (refs) [reply pkg/server/http/server.go]: thanks! strange my linter didn't catch this

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/cache/cache_test.go  6
- pkg/cache/cache.go  4
- pkg/cache/option.go  2
- pkg/middleware/openidconnect.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-proxy/pullRequest?prid=5448512&bid=17866082)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/middleware/openidconnect.go  1
- pkg/cache/option.go  2
- pkg/cache/cache.go  4
- pkg/cache/cache_test.go  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-proxy/pullRequest?prid=5448512&bid=17866082)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/middleware/openidconnect.go  2
- pkg/cache/cache_test.go  6
- pkg/cache/cache.go  4
- pkg/cache/option.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-proxy/pullRequest?prid=5448512&bid=17866082)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/middleware/openidconnect.go  1
- pkg/cache/cache.go  4
- pkg/cache/option.go  2
- pkg/middleware/account_uuid.go  6
- pkg/cache/cache_test.go  6
         

Clones added
============
- pkg/middleware/openidconnect.go  1
- pkg/middleware/account_uuid.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-proxy/pullRequest?prid=5448512&bid=17866082)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/middleware/openidconnect.go  1
- pkg/middleware/account_uuid.go  6
- pkg/cache/cache.go  4
- pkg/cache/cache_test.go  6
- pkg/cache/option.go  2
         

Clones added
============
- pkg/middleware/openidconnect.go  1
- pkg/middleware/account_uuid.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-proxy/pullRequest?prid=5448512&bid=17866082)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/cache/cache_test.go  6
- pkg/cache/cache.go  4
- pkg/middleware/account_uuid.go  6
- pkg/cache/option.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-proxy/pullRequest?prid=5448512&bid=17866082)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/middleware/account_uuid.go  9
- pkg/cache/option.go  2
- pkg/cache/cache_test.go  6
- pkg/cache/cache.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-proxy/pullRequest?prid=5448512&bid=17866082)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/cache/option.go  2
- pkg/middleware/account_uuid.go  9
- pkg/cache/cache.go  4
- pkg/cache/cache_test.go  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-proxy/pullRequest?prid=5448512&bid=17866082)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/cache/option.go  2
- pkg/cache/cache_test.go  6
- pkg/middleware/account_uuid.go  9
- pkg/cache/cache.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-proxy/pullRequest?prid=5448512&bid=17866082)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/middleware/account_uuid.go  9
- pkg/cache/cache.go  4
- pkg/cache/cache_test.go  6
- pkg/cache/option.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-proxy/pullRequest?prid=5448512&bid=17866082)

**Review by kulmann — APPROVED** [review]

**Pull request merged**